errors.php

```diff
@@ -0,0 +1,173 @@
+use search_elastic\local\service\error_service;
+use search_elastic\local\table\errors_table;
+
+defined('MOODLE_INTERNAL') || die();
+
+admin_externalpage_setup('search_elastic_errors');
+require_capability('moodle/site:config', context_system::instance());
+
+// Get parameters.
+$action = optional_param('action', '', PARAM_ALPHAEXT);
+$errorid = optional_param('id', 0, PARAM_INT);
+$confirm = optional_param('confirm', 0, PARAM_BOOL);
+$perpage = optional_param('perpage', 30, PARAM_INT);
+$download = optional_param('download', '', PARAM_ALPHA);
+
+// Search parameters.
+$searchparams = [
+    'docid' => optional_param('docid', '', PARAM_TEXT),
+    'areaid' => optional_param('areaid', '', PARAM_TEXT),
+    'errortype' => optional_param('errortype', '', PARAM_ALPHA),
+    'status' => optional_param('status', '', PARAM_ALPHA),
+    'message' => optional_param('message', '', PARAM_TEXT),
+];
+
+$pageurl = new moodle_url('/search/engine/elastic/errors.php');
+
+$searchform = new error_search_form();
+if ($searchform->is_cancelled()) {
+    redirect(new moodle_url($pageurl));
+} else if ($search_data = $searchform->get_data()) {
+    if (isset($search_data->clear)) {
+        redirect(new moodle_url($pageurl));
+    }
+
+    // Process search filters.
+    $searchparams = error_search_form::process_form_data($search_data);
+
+    // Redirect with search parameters to clean URL.
+    $url = new moodle_url($pageurl, $searchparams);
+    redirect($url);
+}
+
+// Set form data from URL parameters.
+$searchform->set_data((object)$searchparams);
+
+// Clean empty search parameters for table.
+$filters = array_filter($searchparams, function($value) {
+    return $value !== '' && $value !== null && $value !== 0;
+});
+
+$PAGE->set_url(new moodle_url('/search/engine/elastic/errors.php', $filters));
+
+// Get error counts for summary.
+$pendingcount = error_service::get_error_count_by_status(error::STATUS_PENDING);
+$failedcount = error_service::get_error_count_by_status(error::STATUS_FAILED);
+$resolvedcount = error_service::get_error_count_by_status(error::STATUS_RESOLVED);
+
+if ($action && confirm_sesskey()) {
+    if (!$confirm && $action == 'retry_all_errors') {
+        if ($pendingcount == 0 && $failedcount == 0) {
+            notification::info(get_string('noerrors', 'search_elastic'));
+            redirect($PAGE->url);
+        }
+
+        $confirmurl = new moodle_url($PAGE->url, [
+            'action' => 'retry_all_errors',
+            'confirm' => 1,
+            'sesskey' => sesskey(),
+        ]);
+
+        echo $OUTPUT->header();
+        echo $OUTPUT->heading(get_string('indexingerrors', 'search_elastic'));
+        echo $OUTPUT->confirm(get_string('confirm:retryallerrors', 'search_elastic'), $confirmurl, $PAGE->url);
+        echo $OUTPUT->footer();
+        exit();
+    }
+
+    $handler = new error_action_handler();
+    $result = $handler->handle_action($action, $errorid);
+    if ($result['redirect']) {
```

Comment:
It feels like handle_actionalways returns ['redirect']. Maybe it's better to redirect inside error_action_handler class?